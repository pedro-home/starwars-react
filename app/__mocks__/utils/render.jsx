import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import configureStore from 'store';
import Refs from 'contexts/Refs';
import { DEFAULT_REFS } from 'contexts/Refs/constants';

export function renderWithRedux(component, initialState = {}) {
  const store = configureStore(initialState);
  return {
    ...render(<Provider store={store}>{component}</Provider>),
    store,
  };
}

export function renderWithApp(component, initialState = {}, refsValue = DEFAULT_REFS) {
  return renderWithRedux(<Refs.Provider value={refsValue}>{component}</Refs.Provider>, initialState);
}
