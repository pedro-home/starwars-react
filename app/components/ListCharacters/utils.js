import femaleLogo from './assets/female.jpg';
import maleLogo from './assets/male.jpg';
import hermaphroditeLogo from './assets/hermaphrodite.jpg';
import droidLogo from './assets/droid.jpg';
import { OFFSET_SCROLL } from './constants';

export const getCharacterImage = name => {
  let logo;
  switch (name) {
    case 'female':
      logo = femaleLogo;
      break;
    case 'male':
      logo = maleLogo;
      break;
    case 'hermaphrodite':
      logo = hermaphroditeLogo;
      break;
    default:
      logo = droidLogo;
  }

  return logo;
};

export const hasEndedScroll = (offset = OFFSET_SCROLL) => {
  return window.scrollY + window.innerHeight >= document.body.scrollHeight - offset;
};
