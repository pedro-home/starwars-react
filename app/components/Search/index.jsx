import React, { useState, useContext } from 'react';
import { Input, InputGroup, Icon } from 'rsuite';
import PropTypes from 'prop-types';
import Refs from 'contexts/Refs';

const Component = ({ fetch, reset }) => {
  const [value, setValue] = useState('');
  const { results } = useContext(Refs);

  return (
    <InputGroup inside className="search">
      <Input
        placeholder="Search databank..."
        onChange={value => {
          setValue(value);
        }}
      />
      <InputGroup.Button
        onClick={() => {
          reset();

          if (results) {
            window.scrollTo({
              top: results.offsetTop,
              behavior: 'smooth',
            });

            fetch({ search: value });
          }
        }}
      >
        <Icon icon="search" />
      </InputGroup.Button>
    </InputGroup>
  );
};

Component.defaultProps = {
  reset: () => {},
};

Component.propTypes = {
  fetch: PropTypes.func.isRequired,
  reset: PropTypes.func,
};

export default Component;
