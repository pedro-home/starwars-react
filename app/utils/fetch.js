function fetch(url, options) {
  return window.fetch(url, options).then(response => {
    if (response.ok) {
      let body;
      const contentType = response.headers.get('content-type');
      if (contentType && contentType.indexOf('application/json') !== -1) {
        body = response.json();
      } else {
        body = response.text();
      }

      return body;
    }

    throw new Error(`${response.status}: ${response.statusText}`);
  });
}

export default fetch;
