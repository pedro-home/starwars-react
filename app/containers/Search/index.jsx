import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { fetchCharacters, resetCharacters } from 'providers/StarWars/actions';
import Search from 'components/Search';

const Container = ({ fetch, reset }) => {
  return <Search fetch={fetch} reset={reset} />;
};

Container.propTypes = {
  reset: PropTypes.func.isRequired,
  fetch: PropTypes.func.isRequired,
};

function mapDispatchToProps(dispatch) {
  return {
    fetch: filter => dispatch(fetchCharacters(filter)),
    reset: () => dispatch(resetCharacters()),
  };
}

export default connect(null, mapDispatchToProps)(Container);
