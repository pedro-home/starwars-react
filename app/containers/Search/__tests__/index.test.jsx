import React from 'react';
import { renderWithApp } from '__mocks__/utils/render';

import Search from '../index';

describe('Given <Search />', () => {
  let component;

  describe('When initializes', () => {
    beforeEach(() => {
      const { container } = renderWithApp(<Search />);
      component = container;
    });

    it('Then should match the snapshot', () => {
      expect(component).toMatchSnapshot();
    });
  });
});
