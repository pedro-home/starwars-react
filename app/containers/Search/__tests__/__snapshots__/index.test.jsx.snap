// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given <Search /> When initializes Then should match the snapshot 1`] = `
<div>
  <div
    class="rs-input-group search rs-input-group-inside"
  >
    <input
      class="rs-input"
      placeholder="Search databank..."
      type="text"
      value=""
    />
    <a
      class="rs-btn rs-btn-default rs-input-group-btn"
    >
      <i
        class="rs-icon rs-icon-search"
      />
      <span
        class="rs-ripple-pond"
      >
        <span
          class="rs-ripple"
        />
      </span>
    </a>
  </div>
</div>
`;
